### ISSUE TITLE
[Feature] How to support dimension lookup query for transform

### ISSUE THREAD

**Issue by hailin0** [apache/seatunnel#7937]
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

How to support dimension lookup query for transform

And as EtLT data synchronization, I think we need to support lookup query for transform, and I want to hear from you all how you think it can be implemented in SeaTunnel.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by zhdech** [comment]
Implementation method:
Option 1 for batch strategy: When starting the strategy, connect to the database, query the database, and save the data results to the cache; When each line of data arrives, search for cached data.
Option 2 for streaming media strategy: On the basis of 1, regularly refresh cached data to ensure that cached data is updated regularly

**Comment by davidzollo** [comment]
My suggestion is that we can extend sql transform, for example,
```
source {
  MySQL-1 {  // Business Table
    url = "jdbc:mysql://localhost:3306/test"
    driver = "com.mysql.cj.jdbc.Driver" 
    user = "root"
    password = "123456"
    table = "orders"
    plugin_output = "order_source"  
  }

  MySQL-2 {  // Dimension Table
    url = "jdbc:mysql://localhost:3306/test"
    driver = "com.mysql.cj.jdbc.Driver"
    user = "root" 
    password = "123456"
    table = "customers"
    plugin_output = "customers_source"
  }
}

transform {
  Sql {
    plugin_input = ["order_source", "customers_source"] 
    query = """
      SELECT 
        o.*,
        c.name as customer_name,
        c.email as customer_email
      FROM order_source o
      LEFT JOIN customers_source c ON o.customer_id = c.customer_id  
    """
    plugin_output = "order_with_customer"
  }
}

sink {
  jdbc {
        url = "jdbc:postgresql://localhost:5432/test"
        driver = "org.postgresql.Driver"
        user = "admin"
        password = "admin123"
        
        generate_sink_sql = true
        database = "test"
        table = "public.${table_name}"
        schema_save_mode = "CREATE_SCHEMA_WHEN_NOT_EXIST"
        data_save_mode = "APPEND_DATA"
        # enable_upsert = false
    }
}
```

How do you think? @zhdech

**Comment by davidzollo** [comment]
Is there anybody who'd like to implement it?

**Comment by cdgh2021** [comment]
Hi team, is the dimension lookup (lookup join) transform currently implemented？ @davidzollo

**Comment by davidzollo** [comment]
> Hi team, is the dimension lookup (lookup join) transform currently implemented？ [@davidzollo](https://github.com/davidzollo)

Not support for now, would you like to implement it?

**Comment by DanielLeens** [comment]
Thanks for opening this idea.

From the current architecture, SeaTunnel does not have a built-in general-purpose transform-side dimension lookup operator on `dev` yet.

The main blocker is that “dimension lookup” can mean very different things: batch join, streaming lookup with cache, async key-value enrichment, retry semantics, staleness/TTL rules, etc.

Before this can move toward implementation, we probably need one concrete target scenario first.

If you are still following this issue, could you clarify which case you mean:

1. batch lookup against a static dimension table
2. streaming lookup against an external database/cache
3. cache + TTL based online enrichment

With that narrowed down, it will be much easier to decide whether this belongs in SQL transform, a new transform, or a source/sink side capability.

**Comment by davidzollo** [comment]
I'd like to have a try to implement this feature. I'll submit a STIP for this feature!

**Comment by DanielLeens** [comment]
That is the right next step.

A general "dimension lookup" capability is too broad to move straight into implementation, because batch lookup, streaming cache refresh, async key-value enrichment, and consistency/TTL semantics all lead to different runtime contracts.

Please open the STIP first, and keep the first version narrow. The most important things to make explicit are:
1. the concrete target scenario
2. whether this is for batch, streaming, or both
3. cache/refresh and staleness semantics
4. failure and retry behavior
5. whether the capability belongs in SQL, a new transform, or another layer

Until that STIP exists, I would keep this issue as the umbrella feature request rather than starting implementation work directly.

**Comment by DanielLeens** [comment]
## Dynamic Lookup STIP candidate and Phase 0 boundary

I am posting the reviewed architecture candidate here so that the umbrella request has a concrete proposal to discuss. This comment does not claim an assigned STIP number or formal Apache approval.

**Current status:** Proposed architecture candidate. Only an M0 / Phase 0 infrastructure spike is ready to be explored.

**Not approved or implemented:** the M1 lookup baseline, Experimental or GA support, production use, or an Accepted/Final architecture decision.

Architecture snapshot SHA-256: `81465412d02651e88e3ea5f763d8eb285369b82e422c7ade82f87ab6b3403ff4`.

### Narrow target scenario

The proposal is intentionally narrower than a general join engine:

- streaming dimension enrichment only;
- one append-only Fact input, with Kafka as the first capability target;
- one single-table Dimension bootstrap/changelog input, with MySQL CDC as the first capability target;
- primary-key equality lookup;
- latest processing state for future Fact records, without retroactive correction of already emitted results;
- same-parallelism restore in the first version;
- primary-key changes, unsupported schema changes, shared source gates, and unsupported topologies fail fast.

The two explicit inputs are:

- port `0`: Fact;
- port `1`: Dimension bootstrap/changelog.

This is not a general relational join, union, deduplication, windowing, temporal-table engine, or arbitrary multi-stream computation proposal. It also does not reuse the existing single-input Transform contract, because the two inputs have different schemas, lifecycle rules, checkpoint ownership, and recovery order.

### Required engine architecture

The current logical graph can describe multiple upstream actions, but the execution path does not preserve a true multi-input operator. A multi-input target is split into source-rooted pipelines, and the physical flow is rebuilt once per source. That duplicates the target action and loses input-port and channel identity.

The proposed chain is:

```text
DynamicLookupAction
  -> PortAwareLogicalEdge
  -> PortAwareExecutionEdge
  -> one Pipeline target vertex
  -> one physical multi-input target per target subtask
  -> deployment input-port/channel descriptors
  -> MultiInputTask
```

The same `operatorUid` also owns exactly one control-plane chain:

```text
DynamicLookupAction
  -> DynamicLookupCoordinatorAction
  -> DynamicLookupCoordinatorTask
  -> CoordinatorStateKey(operatorUid)
  -> Physical / Deployment / Checkpoint plans
```

The complete architecture additionally defines versioned channel identities, attempt fencing, per-channel barriers, managed state handles, checkpoint terminal CAS, source restore gates, bootstrap ordering, and bounded resource admission. Those protocols are later phase gates and are not being claimed by the first Draft PR.

### Phase 0 sequence

Phase 0 must remain serial and independently reviewable:

1. PR-1: port-aware multi-input physical topology;
2. PR-2: HASH exchange and canonical key encoding;
3. PR-3: per-channel barrier/alignment state machine;
4. PR-4: managed state, handles, restore, and checkpoint terminal lifecycle;
5. PR-5: dimension bootstrap, Fact gate, leases, and attempt fencing.

M1 lookup semantics cannot start until those gates, compatibility fixtures, fault injection, and resource bounds have evidence and receive another maintainer review.

### First Draft PR scope

The first Draft PR is limited to a non-user-reachable PR-1 planning/deployment skeleton:

1. add a versioned port-aware logical edge under a new serialization class ID;
2. preserve edge identity, target input port, and the versioned exchange envelope through Logical, Execution, Pipeline, Physical, Deployment, and Checkpoint plans;
3. keep one target action/vertex instead of duplicating it for each source root;
4. create immutable logical and physical channel descriptors;
5. create one multi-input task shell per target subtask and fail fast if someone tries to run it before the later protocols exist;
6. create one operator-scoped coordinator task and checkpoint key;
7. keep existing single-input, Union, multiple-Source, and multiple-Sink graphs on their legacy path.

PR-1 does **not** add user-facing Dynamic Lookup syntax and does not claim working lookup execution. It does not implement HASH routing, business-record transport, barrier alignment, managed lookup state, source restore gates, bootstrap commands, or production fencing.

### Compatibility invariants

- The existing `LogicalEdge` class ID and exact two-`long` payload remain frozen.
- The new edge uses a separate tagged class ID and explicit format version.
- Legacy and port-aware edges are symmetrically unequal.
- New edge identity includes edge ID, both endpoints, target input port, and canonical exchange bytes.
- Reusing an edge ID with a different descriptor fails fast.
- Pure legacy graphs retain their current pipeline split and task topology.
- A mixed-version deployment capability handshake is still an open PR-1 gate and will not be represented as complete by DTOs alone.

### Requested review

The initial review should focus on whether this narrow scenario and Phase 0 split are acceptable, especially:

- whether the feature must remain separate from general Transform and SQL join semantics;
- whether the old `LogicalEdge` compatibility boundary is sufficient;
- whether PR-1 should be split further before real deployment-attempt fencing is added;
- whether the Fact/Dimension source ownership restrictions are acceptable for the first version.

The implementation PR will remain Draft and will use `Refs #7937`; it will not close this umbrella issue.

**Comment by DanielLeens** [comment]
Draft implementation spike opened: #11559

This PR implements only the **Phase-0 / M0 terminal planning and deployment-descriptor slice** of the proposal above. It is deliberately non-deployable and non-runnable, and both physical-plan initialization and direct start fail before any source or lookup task can run.

It must not be treated as PR-1 complete or merge-ready. The remaining completion gates are:

1. real exchange, barrier, state, restore/gate, bootstrap, output, and lookup row semantics;
2. scheduler-owned deployment-attempt identity plus transport fencing;
3. mixed-version JobMaster/Worker capability handshake;
4. the full Gate A multi-channel, different-parallelism, legacy-regression, deployment, and checkpoint runtime matrix.

The Draft body records the exact static validation performed and the local compile/test/E2E evidence that is still absent.

**Comment by DanielLeens** [comment]
Dynamic Lookup implementation draft has been pushed to PR #11559.

PR: https://github.com/apache/seatunnel/pull/11559
Head commit: `183bfbcc98aa3f5061ee8df7968d3f1c1a2e3d4a`

Design-to-implementation summary:

1. Dedicated action and parser
   - Adds `dynamic_lookup` parser support.
   - Builds a dedicated `DynamicLookupAction` instead of modeling lookup as a transform.
   - Produces an output `CatalogTable` from the declared fact/dimension projection.

2. Fact source gate and restore ownership
   - Adds `FactSourceGateCapability`, `SourceGateState`, and `SourceGateCommand` API contracts.
   - Kafka source now advertises `FACT_SOURCE_GATE_V1` and stages splits while the fact gate is closed.
   - Closed-gate Kafka split ownership is snapshotted through native `snapshotState`, so restore keeps the source as the split owner.
   - Fact split activation happens only after the lookup side sends an explicit `OPEN` command.

3. Bootstrap and fact activation ordering
   - Dynamic lookup consumes dimension and fact ports from same-task-group blocking queues for this draft.
   - Dimension state is checkpointed first.
   - The first fact-gate `OPEN` is sent only from `notifyCheckpointComplete` for the checkpoint that contains the dynamic lookup dimension state.
   - If restoring from a durable lookup checkpoint, the fact gate opens during runtime open because the committed checkpoint already implies `FACT_POSITIONS_DURABLE`.

4. Checkpoint persistence and durable phase metadata
   - Adds `CheckpointIntent` to `CompletedCheckpoint`.
   - Dynamic lookup checkpoint contents are detected by state envelope magic.
   - Completed checkpoints carrying lookup state get purpose `DYNAMIC_LOOKUP_FACT_POSITION_ANCHOR`, target phase `FACT_POSITIONS_DURABLE`, and a SHA-256 digest over canonical action state bytes.
   - Completed checkpoint payloads are encoded with magic/version/payload length/payload/SHA-256 and strict legacy fallback.

5. Runtime lookup semantics
   - Maintains dimension state keyed by planner-resolved key fields.
   - Supports LEFT and INNER lookup projection.
   - Fact input is append-only in M1.
   - Dimension UPDATE_BEFORE/UPDATE_AFTER must be adjacent and same-key; different keys fail fast as primary-key updates.
   - Schema change events fail fast.
   - Checkpoint barriers align across fact and dimension ports before snapshot and downstream forwarding.

6. Channel and identity groundwork
   - Logical channel keys have canonical bytes and SHA-256 digest.
   - Channel attempt identity includes job execution epoch.
   - Channel envelopes include canonical digest and duplicate/conflict detection.
   - The current draft runtime uses same-task-group blocking queues; remote channel transport wiring remains a follow-up evidence gate before leaving Draft.

7. Resource admission
   - Parser enforces M1 mode/resource constraints:
     - disk-backed state
     - no TTL
     - one concurrent snapshot
     - max 4 GiB logical state per subtask
     - max 512 MiB resident state per subtask
   - It also checks the uncommitted snapshot count/byte formulas, cleanup throughput, local disk reservation, and remote staging quota before accepting the plan.

Local checks performed:

- `git diff --check`
- `git diff --cached --check`
- AOSP google-java-format apply/dry-run on the known changed Java files

Not run locally by SeaTunnel local policy:

- Maven compile
- unit tests
- integration tests
- E2E tests
- Docker/service startup
- real SeaTunnel job execution

The PR is intentionally still Draft. Current-head GitHub CI and maintainer review remain the compile/runtime evidence gate.

**Comment by DanielLeens** [comment]
Update for the dynamic lookup implementation draft:

PR #11559 has been updated to head `5f6d15218e72f4522e20e33b015ce998b73ac201`.

This head includes the complete M1 draft implementation path described by the current architecture draft:

- top-level `dynamic_lookup` parser support and `DynamicLookupAction` materialization
- same-task-group two-input dynamic lookup runtime
- LEFT/INNER projection, append-only fact enforcement, dimension same-key update-pair enforcement, and schema-change fail-fast handling
- per-port checkpoint barrier alignment before snapshot/ack/emit
- Kafka fact source gate staging/restore/open support
- dynamic lookup checkpoint intent metadata and versioned checkpoint payload codec with legacy fallback
- channel identity/digest primitives for the later remote-channel path
- CDC incremental source capability declarations
- en/zh architecture documentation for the user-visible `dynamic_lookup` configuration and M1 limits

A maintainer recheck for the previously reported barrier-alignment P0 passed on the latest implementation head. The PR intentionally remains Draft because local policy only allows formatting/static checks for SeaTunnel; GitHub CI and reviewer feedback remain the compile/runtime evidence gate.

**Comment by DanielLeens** [comment]
Correction for the previous update comment: the actual latest PR #11559 head is `5f6d15218e61d7b915ae4c961001d8b38dec5d22`. The abbreviated commit remains `5f6d15218e`; the implementation summary and PR link are unchanged.
